File: app/src/main/java/de/g00fy2/developerwidget/WidgetConfigActivity.java
package de.g00fy2.developerwidget;

import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;
import android.widget.RemoteViews;
import java.lang.reflect.Field;

public class WidgetConfigActivity extends AppCompatActivity {

  private int widgetId;
  private RemoteViews remoteViews;
  private Context context;

  @Override public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_widget_config);

    setResult(RESULT_CANCELED);
    context = this;
    remoteViews = new RemoteViews(context.getPackageName(), R.layout.appwidget_layout);

    Intent intent = getIntent();
    Bundle extras = intent.getExtras();
    if (extras != null) {
      widgetId =
          extras.getInt(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID);
    }

    if (widgetId == AppWidgetManager.INVALID_APPWIDGET_ID) {
      finish();
    }

    Button apply = (Button) findViewById(R.id.apply);
    apply.setOnClickListener(new View.OnClickListener() {
      @Override public void onClick(View v) {
        setWidgetText();

        AppWidgetManager.getInstance(context).updateAppWidget(widgetId, remoteViews);

        Intent resultValue = new Intent();
        resultValue.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, widgetId);
        setResult(RESULT_OK, resultValue);
        finish();
      }
    });
  }

  private void setWidgetText() {
    remoteViews.setTextViewText(R.id.device_info, getDeviceInfo());
    remoteViews.setTextViewText(R.id.release, getVersionAndCodename());
    remoteViews.setTextViewText(R.id.sdk_int, getSDKVersion());
  }

  private String getDeviceInfo() {
    return Build.MANUFACTURER + " " + Build.MODEL;
  }

  private String getVersionAndCodename() {
    return "Android " + Build.VERSION.RELEASE + " (" + getAndroidCodename() + ")";
  }

  private String getAndroidCodename() {
    Field[] fields = Build.VERSION_CODES.class.getFields();
    return fields[Build.VERSION.SDK_INT + 1].getName();
  }

  private String getSDKVersion() {
    return "API " + Integer.toString(Build.VERSION.SDK_INT);
  }
}

File: app/src/main/kotlin/de/g00fy2/developerwidget/activities/widget/WidgetConfigActivity.kt
package de.g00fy2.developerwidget.activities.widget

import android.app.Activity
import android.appwidget.AppWidgetManager
import android.content.Intent
import android.os.Build.VERSION
import android.os.Build.VERSION_CODES
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.webkit.WebView
import androidx.annotation.ContentView
import androidx.core.content.ContextCompat.startActivity
import androidx.recyclerview.widget.LinearLayoutManager
import de.g00fy2.developerwidget.R
import de.g00fy2.developerwidget.activities.about.AboutActivity
import de.g00fy2.developerwidget.base.BaseActivity
import de.g00fy2.developerwidget.data.DeviceDataItem
import de.g00fy2.developerwidget.data.DeviceDataProvider
import kotlinx.android.synthetic.main.activity_widget_config.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

@ContentView(R.layout.activity_widget_config)
class WidgetConfigActivity : BaseActivity(), WidgetConfigContract.WidgetConfigView {

  private var updateExistingWidget = false
  private var widgetId: Int = AppWidgetManager.INVALID_APPWIDGET_ID
  private lateinit var adapter: DeviceDataAdapter

  public override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setResult(Activity.RESULT_CANCELED)

    intent.extras?.let {
      widgetId = it.getInt(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID)
      updateExistingWidget = it.getBoolean(EXTRA_APPWIDGET_UPDATE_EXISTING)
    }

    setActionbarElevationListener(widget_config_root_scrollview)

    adapter = DeviceDataAdapter()
    recyclerview.setHasFixedSize(true)
    recyclerview.layoutManager = LinearLayoutManager(this)
    recyclerview.isNestedScrollingEnabled = false
    recyclerview.adapter = adapter

    // set up webview pre oreo to get implementation information
    if (VERSION.SDK_INT in VERSION_CODES.LOLLIPOP until VERSION_CODES.O) {
      WebView(this)
    }

    if (widgetId == AppWidgetManager.INVALID_APPWIDGET_ID) {
      finish()
      return
    }

    apply_button.setOnClickListener {
      val appWidgetManager = AppWidgetManager.getInstance(this)
      WidgetProvider.updateWidget(this, appWidgetManager, widgetId)

      val resultValue = Intent()
      resultValue.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, widgetId)
      setResult(Activity.RESULT_OK, resultValue)
      finish()
    }
    if (updateExistingWidget) apply_button.setText(R.string.update_widget)
  }

  override fun onResume() {
    super.onResume()
    launch {
      getDeviceData().let {
        setWidgetFields(it.toMap())
        adapter.notifyDataSetChanged() // workaround to show items
        adapter.submitList(it)
      }
    }
  }

  override fun onCreateOptionsMenu(menu: Menu): Boolean {
    menuInflater.inflate(R.menu.configuration_menu, menu)
    return true
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    return when (item.itemId) {
      R.id.about_button -> {
        val intent = Intent(this, AboutActivity::class.java)
        startActivity(intent)
        true
      }
      else -> super.onOptionsItemSelected(item)
    }
  }

  private fun setWidgetFields(data: Map<String, DeviceDataItem>) {
    device_title_textview.text = data[DeviceDataProvider.DEVICE_NAME]?.value ?: ""
    var subtitle = data[DeviceDataProvider.RELEASE]?.let { getString(it.title) + " " + it.value + " | " } ?: ""
    subtitle += data[DeviceDataProvider.SDK]?.let { getString(it.title) + " " + it.value }
    device_subtitle_textview.text = subtitle
  }

  private suspend fun getDeviceData(): List<Pair<String, DeviceDataItem>> {
    return withContext(Dispatchers.IO) {
      DeviceDataProvider
        .getStaticDeviceData()
        .plus(DeviceDataProvider.getHardwareData(this@WidgetConfigActivity))
        .plus(DeviceDataProvider.getSoftwareInfo(this@WidgetConfigActivity))
        .plus(DeviceDataProvider.getHeaderItems())
        .toList()
        .filter { (_, value) -> value.value.isNotBlank() || value.isHeader }
        .sortedWith(
          compareBy(
            { it.second.category.ordinal },
            { !it.second.isHeader },
            { getString(it.second.title) })
        )
    }
  }

  companion object {
    const val EXTRA_APPWIDGET_UPDATE_EXISTING = "updateExistingWidget"
  }
}